### PR TITLE
[Bugfix] Transformers backend: apply learned lm_head.bias for tied-embedding models

### PR DESCRIPTION
**Pull request by JohnLangford**
## The bug

vLLM's Transformers backend (`CausalMixin`, `vllm/model_executor/models/transformers/causal.py`) silently drops a learned `lm_head.bias` for tied-embedding models, producing garbage greedy output. Three independent failure points, all reached when `tie_word_embeddings=True`:

- **(A)** `self.skip_prefixes.append("lm_head.")` — `AutoWeightsLoader._can_skip` matches by `startswith`, so this drops `lm_head.bias` along with the (correctly skipped, tied) `lm_head.weight`.
- **(B)** `ParallelLMHead(...)` is built with no `bias=`, so it registers `bias=None` and there is no slot to load a bias into.
- **(C)** `compute_logits` calls `self.logits_processor(self.lm_head, hidden_states)` with no bias. `LogitsProcessor._get_logits` → `UnquantizedEmbeddingMethod.apply(layer, x, bias=embedding_bias)` uses only the **passed** bias and never reads `layer.bias`. So even a populated `lm_head.bias` parameter would be ignored.

The HF reference computes `logits = hidden @ Wᵀ + b`; vLLM computes `logits = hidden @ Wᵀ`. The dropped per-vocab bias (e.g. std ~1.1 / range ~6.6 logits on the checkpoint that surfaced this) is on the order of inter-token logit gaps, so it reorders the greedy argmax → cascading garbage. (`logit_softcap` is a monotonic `tanh`; it preserves argmax and is not involved.)

Reproduced on current `main` (and on the `v0.22.0` line) with a tied-embedding checkpoint that ships a learned `lm_head.bias`, served via `--model-impl transformers`.

## The fix

`causal.py` only:
1. Detect a present `lm_head.bias` (`_has_lm_head_bias`: known config-flag aliases, else the model's `lm_head` module).
2. When tied, skip only `lm_head.weight` (not the whole prefix) so the bias still loads.
3. Build `ParallelLMHead(bias=has_lm_head_bias)` so there is a slot to load into.
4. Forward `self.lm_head.bias` into `LogitsProcessor`'s existing `embedding_bias` arg in `compute_logits`.

`tie_weights` only reassigns `.weight`, so a tied weight and an independent bias compose cleanly. When `lm_head.bias` is absent everything is a no-op (identical to today). This also fixes the latent untied-embeddings-with-`lm_head.bias` case, since the head's `bias=` and the `compute_logits` pass-through no longer depend on the tie flag. No public API change — `embedding_bias` is an existing `LogitsProcessor` parameter.

## Tests

`tests/models/test_transformers.py` (GPU-free):
- `test_causal_mixin_compute_logits_forwards_lm_head_bias` — spies the `LogitsProcessor` and asserts `compute_logits` forwards `lm_head.bias` (guards (C), the load-bearing change).
- `test_causal_mixin_has_lm_head_bias_detection` — the new detector across config-flag aliases and the module fallback.

A tied-embedding model with **no** `lm_head.bias` (the common case) is unaffected.

### CODE REVIEW

**Comment by hmellor** [comment]
Can you please give an example where the weight would be tied but the bias would not?

File: vllm/model_executor/models/transformers/causal.py [hunk 1]
```diff
@@ -61,6 +68,22 @@ def __init__(self, *, vllm_config: "VllmConfig", prefix: str = ""):
...
+                return True
+        lm_head = getattr(self.model, "lm_head", None)
+        return getattr(lm_head, "bias", None) is not None
+
     def compute_logits(self, hidden_states: "torch.Tensor") -> "torch.Tensor | None":
-        logits = self.logits_processor(self.lm_head, hidden_states)
+        # Apply the learned output bias (if any). The unquantized logits path
+        # only honours a bias passed here, never ``lm_head.bias`` directly.
+        bias = getattr(self.lm_head, "bias", None)
+        logits = self.logits_processor(self.lm_head, hidden_states, bias)
```
Comment (hmellor) [inline vllm/model_executor/models/transformers/causal.py]: `ParallelLMHead` always registers `self.bias`, so we don't need to use `getattr`

```suggestion
        logits = self.logits_processor(self.lm_head, hidden_states, self.lm_head.bias)
```

File: vllm/model_executor/models/transformers/causal.py [hunk 2]
```diff
@@ -61,6 +68,22 @@ def __init__(self, *, vllm_config: "VllmConfig", prefix: str = ""):
...
+
+        HF exporters that emit a learned output bias set an explicit config
+        flag; some configs spell it differently, so accept any known alias and
+        fall back to the model's own ``lm_head`` module when present.
+        """
+        for attr in ("lm_head_bias", "use_lm_head_bias", "output_bias"):
+            if getattr(self.text_config, attr, None):
+                return True
+        lm_head = getattr(self.model, "lm_head", None)
+        return getattr(lm_head, "bias", None) is not None
```
Comment (hmellor) [inline vllm/model_executor/models/transformers/causal.py]: This will never be true because `self.model` is the base model (i.e. no lm head by definition)

**Review by hmellor — COMMENTED** [review]
This is a nice edge case you've found! Thank you for the fix.

Unfortunately I don't think the bias detection is robust enough as it is now. If you could give an example checkpoint where this is an issue I'll see if we can find biases 100% of the time

**Comment by JohnLangford** [comment]
> Can you please give an example where the weight would be tied but the bias would not?

There is no reason to add a bias term to the initial embedding process itself since you want the embedding to be all about information concerning the tokens.  And of course you want to tie things if you are minimizing the number of parameters in a parameter constrained setting.  So, the question is really: why do you want to have a bias on the output?  My experience is that the output settles into a pattern where it has a 'prior' over tokens regardless of whether or not you have a bias term.  Without the bias, that has to be manufactured with a residual which (a) occupies the residual and (b) requires a very large scale prenorm to make it near-constant.  This stresses the network, which seems undesirable, with the obvious solution being a bias term in the output head.

I am actively using this.

**Comment by JohnLangford** [comment]
> This is a nice edge case you've found! Thank you for the fix.
> 
> Unfortunately I don't think the bias detection is robust enough as it is now. If you could give an example checkpoint where this is an issue I'll see if we can find biases 100% of the time

The model that I'm working on is presently private, unfortunately.

File: vllm/model_executor/models/transformers/causal.py (hunk 2 above)
Comment (JohnLangford) [reply vllm/model_executor/models/transformers/causal.py]: As an alternative, what's the right way to do something like this in vllm?

def _has_lm_head_bias(self) -> bool:
    for attr in ("lm_head_bias", "use_lm_head_bias", "output_bias"):
        if getattr(self.text_config, attr, None):
            return True                                
    return "lm_head.bias" in self._checkpoint_weight_names() # from the index weight_map

**Comment by hmellor** [comment]
Thanks for the updates. That's ok, I can test theoretically and then ask you to check with your private model

**Comment by hmellor** [comment]
@JohnLangford please try what I have just pushed

**Comment by JohnLangford** [comment]
This works, but please check the torch.empty(...) call---I think you need a device=<gpu>.

**Comment by hmellor** [comment]
Great!

Setting the device is not necessary (it wasn't before this PR either) because vLLM models are instantiated with a decorator which changes the default device:

https://github.com/vllm-project/vllm/blob/4559c43a9526597c00cbcc4f59979496500268d1/vllm/model_executor/model_loader/base_loader.py#L47-L59

**Review by hmellor — APPROVED** [review]
